### PR TITLE
fix(examples/fsdp_tp): migrate FSDP1→FSDP2 (fully_shard) — fixes agpt-2b seq=8192 OOM

### PR DESCRIPTION
**Pull request by saforem2**
## Problem

`ezpz.examples.fsdp_tp` OOM'd in `loss.backward()` (`UR_RESULT_ERROR_OUT_OF_RESOURCES`) training **agpt-2b at seq=8192** on XPU — while torchtitan trains the *same 2B model at the same seq length* fine, and the model uses <20% of memory.

## Root cause (systematic debug, experiments A–E on Sunspot PVC)

| exp | config (seq=8192, 12 ranks) | result |
|---|---|---|
| A | eager loss, AC=none | OOM in backward |
| C | eager loss, **seq=2048** | ✓ 39.95 GB peak |
| B | `--compile` | OOM in backward |
| D | chunked loss | OOM in backward |
| E | chunked loss + AC=full | OOM in backward |

Ruled out: the loss transient (chunked loss didn't fix it), activation checkpointing (AC=full didn't fix it), `torch.compile`, and shard count (the user's tp=1 run had dp_shard=48 > torchtitan's 24 and still OOM'd). The one architectural difference: **ezpz `fsdp_tp` used FSDP1 (`FullyShardedDataParallel`); torchtitan uses FSDP2 (`fully_shard`).** FSDP2's per-module sharding — applied separately to the 256K-vocab embedding, each block, and `[norm, output]` — bounds the backward gradient/activation memory where FSDP1's single flat-parameter wrap does not.

## Changes (all in `examples/fsdp_tp.py`)

- **`parallelize()` (TP path):** `FSDP(model, sharding_strategy=…)` → per-module `fully_shard` (embedding / each block / `[norm, output]` / root), following torchtitan's `apply_fsdp`. TP plan unchanged.
- **HF path:** FSDP1 auto-wrap → per-block `fully_shard`.
- `MixedPrecision` → `MixedPrecisionPolicy`.
- **`--sharding-strategy`** now maps to FSDP2 `reshard_after_forward` (`full_shard`→`True`, `shard_grad_op`/`no_shard`→`False`); legacy enum names kept as the CLI surface (no break); help updated.
- **Activation checkpointing** now applied per-block **before** `fully_shard` (torchtitan ordering); the FSDP1 AC-after-wrap order is removed.
- **DTensor-safety in metrics:** `_accumulate_stats` / `_collect_layer_grad_norms` unwrap DTensor params/grads via `.to_local()` before `numel`/`max`/`sum` — FSDP2 params are DTensors whose `.numel()` is global, so empty local shards on some ranks previously raised `max(): Expected reduction dim ...`.
- Drop now-unused FSDP1 imports, `use_orig_params` workarounds, and `functools`.

Also rides along: the **`--loss-impl {eager,chunked,compiled}`** toggle added while diagnosing — a correct, independently useful memory knob (chunked/compiled cross-entropy), though FSDP2 is what actually fixes the OOM.

## Verification (Sunspot, 4 nodes, torch 2.13+xpu)

- **agpt-2b seq=8192 (eager)** → trains to completion (was OOM in all 5 FSDP1 experiments).
- **User's seq=7320 `--loss-impl=compiled --compile`** → trains to completion (was the reported crash).
- **112 `test_fsdp_tp_*` / compile-flag / flops unit tests pass.**

> Benign warnings observed (documented, not bugs): FSDP2 "returned a view tensor" (model uses out-of-place residual adds, so gradients are correct), the `_NormPartial` 2-all-reduce perf hint (SequenceParallel RMSNorm; throughput only), and inductor "complex operators" (RoPE complex math falls back to eager on XPU).

Follow-up: docs PR #172's `fsdp-tp.md` snippet line-ranges + FSDP2 details will need a refresh after this lands.

## Summary by Sourcery

Switch the fsdp_tp example to use FSDP2 per-module sharding to resolve long-sequence OOMs and align behavior with torchtitan, while adding a configurable loss implementation for better memory control.

New Features:
- Introduce configurable cross-entropy loss implementations (`eager`, `chunked`, `compiled`) with an optional chunk size flag to control peak memory usage during backward.
- Add a CLI flag to select the loss implementation and parameterize chunk size for chunked cross-entropy.

Enhancements:
- Update sharding configuration to map legacy FSDP1 strategy names onto FSDP2 `reshard_after_forward` behavior without breaking existing CLI.
- Refactor tensor-parallel `parallelize()` and HF model paths to apply FSDP2 `fully_shard` on embeddings, transformer blocks, output layers, and the root module, with activation checkpointing ordered per-block before sharding.
- Adopt `MixedPrecisionPolicy` for FSDP2 mixed-precision configuration while preserving fp32 debugging mode.
- Improve gradient and metric collection utilities to handle FSDP2 DTensor parameters and grads safely by operating on local shards and skipping empty tensors.

Tests:
- Ensure existing `test_fsdp_tp_*`, compile-flag, and flops unit tests pass with the new FSDP2 and loss implementation changes.

### CODE REVIEW

File: src/ezpz/examples/fsdp_tp.py [hunk 1]
```diff
@@ -1205,7 +1390,14 @@ def _collect_layer_grad_norms(model: nn.Module) -> list[float]:
                 layer_id = int(layer_str)
             except Exception:
                 continue
-            grad = param.grad.float()
+            grad = param.grad
```
[the remark this answers is not in the document]
Comment (saforem2) [reply src/ezpz/examples/fsdp_tp.py:1491]: Already handled — `_collect_layer_grad_norms` opens its loop with `if param.grad is None: continue` (and `.layers.` name filtering) before any `.to_local()/.float()`. This review was against an earlier commit; no change needed.

File: src/ezpz/examples/fsdp_tp.py [hunk 2]
```diff
@@ -312,15 +310,33 @@
...
+# FSDP2 (`fully_shard`) has no `sharding_strategy` enum — sharding behavior
+# is controlled by `reshard_after_forward` (bool). We keep the legacy
+# FSDP1 strategy NAMES as the CLI surface (so existing scripts don't break)
+# and map each to the closest FSDP2 reshard policy:
+#   - full_shard  (ZeRO-3): reshard params after forward -> lowest memory  -> True
+#   - shard_grad_op (ZeRO-2): keep params unsharded after forward          -> False
+#   - no_shard    (ZeRO-0 / replicate): also keep unsharded (closest FSDP2 -> False;
+#       true no-shard = don't shard at all, not expressible per-module here)
+#   - hybrid_shard*: FSDP1 inter/intra-node hybrid has no direct one-flag
+#       FSDP2 equivalent (needs an explicit 2D mesh); map to full reshard.
```
[the remark this answers is not in the document]
Comment (saforem2) [reply src/ezpz/examples/fsdp_tp.py]: Fixed in a79e286. Comment now states hybrid_shard -> reshard-after-forward (True), hybrid_shard_zero2 -> keep-unsharded (False), matching SHARDING_STRATEGIES.

File: src/ezpz/examples/fsdp_tp.py [hunk 3]
```diff
@@ -739,19 +869,17 @@ def parse_args(argv: Optional[list[str]] = None):
...
+            "which controls sharding via `reshard_after_forward` rather than "
+            "a strategy enum; the legacy FSDP1 names below are kept as the "
+            "CLI surface and mapped to the nearest FSDP2 policy. "
+            "`full_shard` (default, ZeRO-3): reshard params after forward — "
+            "lowest memory, params re-all-gathered in backward. "
+            "`shard_grad_op` (ZeRO-2): keep params unsharded after forward — "
+            "more memory, avoids the backward all-gather. "
+            "`no_shard`: mapped to keep-unsharded (FSDP2 has no true "
+            "per-module no-shard). `hybrid_shard*`: FSDP1 hybrid has no "
+            "one-flag FSDP2 equivalent — mapped to full reshard."
```
[the remark this answers is not in the document]
Comment (saforem2) [reply src/ezpz/examples/fsdp_tp.py]: Fixed in a79e286. Help text now distinguishes hybrid_shard (reshard) from hybrid_shard_zero2 (keep-unsharded).

File: src/ezpz/examples/fsdp_tp.py [hunk 4]
```diff
@@ -379,6 +395,120 @@ def _slice_for_sequence_parallel(
...
+    """
+    flat_logits = logits.reshape(-1, logits.size(-1))
+    flat_labels = labels.reshape(-1)
+    n_rows = flat_labels.shape[0]
+
+    valid = (flat_labels != ignore_index).sum()
+    # Avoid div-by-zero on an all-ignored microbatch.
+    denom = valid.clamp(min=1).to(logits.dtype)
+
+    total = logits.new_zeros(())
```
[the remark this answers is not in the document]
Comment (saforem2) [reply src/ezpz/examples/fsdp_tp.py]: Fixed in a79e286. denom is now `.float()` and the accumulator is `torch.zeros((), dtype=torch.float32)`, with `chunk_loss.float()` per add — fp32 throughout to match eager and avoid per-chunk promotion.

File: src/ezpz/examples/fsdp_tp.py [hunk 5]
```diff
@@ -901,16 +1061,26 @@ def parse_args(argv: Optional[list[str]] = None):
...
+    length. Activation checkpointing (when requested) is applied to each
+    block BEFORE ``fully_shard`` so the checkpoint envelope sits inside the
+    sharded unit (torchtitan's ordering).
+    """
     tp_mesh = device_mesh["tp"]
     dp_mesh = device_mesh["dp"]
 
-    if isinstance(sharding_strategy, str):
-        sharding_strategy = SHARDING_STRATEGIES.get(sharding_strategy, None)
+    reshard = _reshard_after_forward(sharding_strategy)
```
[the remark this answers is not in the document]
Comment (saforem2) [reply src/ezpz/examples/fsdp_tp.py:1152]: Fixed in a79e286. Removed the unused device_id parameter from parallelize() and its call site.

File: src/ezpz/examples/fsdp_tp.py [hunk 6]
```diff
@@ -312,15 +310,33 @@
...
+#   - hybrid_shard*: FSDP1 inter/intra-node hybrid has no direct one-flag
+#       FSDP2 equivalent (needs an explicit 2D mesh); map to full reshard.
+# This is the set of valid `--sharding-strategy` values.
 SHARDING_STRATEGIES = {
-    "no_shard": ShardingStrategy.NO_SHARD,
-    "full_shard": ShardingStrategy.FULL_SHARD,
-    "shard_grad_op": ShardingStrategy.SHARD_GRAD_OP,
-    "hybrid_shard": ShardingStrategy.HYBRID_SHARD,
-    "hybrid_shard_zero2": ShardingStrategy._HYBRID_SHARD_ZERO2,
+    "no_shard": False,
```
[the remark this answers is not in the document]
Comment (saforem2) [reply src/ezpz/examples/fsdp_tp.py:328]: Addressed in a79e286 with a runtime warning: when --sharding-strategy=no_shard runs multi-rank, we now warn that FSDP2 still shards params/grads/optimizer state across the dp mesh (the flag only disables post-forward resharding) and point to plain DDP for truly replicated params. A per-module 'skip fully_shard' path isn't expressible in this example's FSDP2 wrap, so we surface it loudly rather than silently mislead.

**Comment by saforem2** [comment]
Addressed automated review feedback in `a79e286`:

- **`hybrid_shard*` docs** (Copilot ×2): corrected both the `SHARDING_STRATEGIES` comment and `--sharding-strategy` help — `hybrid_shard` → reshard-after-forward, `hybrid_shard_zero2` → keep-unsharded.
- **chunked-CE precision** (Copilot): `_cross_entropy_chunked` now accumulates and divides in fp32 (was casting to bf16 logits dtype), matching eager numerics and dropping per-chunk promotion.
- **dead `device_id`** (Copilot): removed the unused parameter from `parallelize()` + call site.
- **`no_shard` semantics** (codex P2): added a multi-rank runtime warning that FSDP2 still shards params/grads/optimizer state (the flag only disables post-forward resharding); pointed to plain DDP for truly replicated params.
- **grad-None guard** (sourcery): already present (`if param.grad is None: continue`) — no change needed.

All threads replied + resolved.
